File: Implementation/Graph/python/dfs.py
def dfsAL(graph, start_node):
    stack = []
    visited = []
    stack.append(start_node)
    while stack:
        node = stack.pop()
        if node not in visited:
            visited.append(node)
            stack.extend(list(map(lambda x: x[0], graph[node])))

    return visited

def dfsAM(graph, start_node):
    stack = []
    visited = []
    stack.append(start_node)
    while stack:
        node = stack.pop()
        if node not in visited:
            visited.append(node)
            for idx, weight in enumerate(graph[node]):
                if weight != 0:
                    stack.append(idx)

    return visited

File: Implementation/Graph/go/internal/graph/dfs.go
package graph

import (
	"fmt"

	"algorithm.implementation/datastruct"
)

func (g AdjacencyList) DfsAL(start string) []string {
	visited := make([]string, 0)
	s := datastruct.Stack{}
	s.Push(start)
	for s.GetLength() != 0 {
		node := fmt.Sprintf("%v", s.Pop())
		var exist bool
		for _, val := range visited {
			exist = val == node || exist
		}

		if exist {
			continue
		}

		visited = append(visited, node)
		for _, val := range g.Graph[node] {
			s.Push(val.name)
		}
	}

	return visited
}

func (g AdjacencyMatrix) DfsAM(start int) []int {
	visited := make([]int, 0)
	s := datastruct.Stack{}
	s.Push(start)
	for s.GetLength() != 0 {
		node := s.Pop().(int)
		var exist bool
		for _, val := range visited {
			exist = val == node || exist
		}

		if exist {
			continue
		}

		visited = append(visited, node)
		for idx, val := range g.Graph[node] {
			if val == 0 {
				continue
			}
			s.Push(idx)
		}
	}

	return visited
}
